File: 02. Algorithms/01. Arrays/22. Max Product Subarray/Solution.java
package com.company;

import java.util.Scanner;

class Solution {

    public static int maxProduct(int[] nums) {
        int[] max = new int[nums.length];
        int[] min = new int[nums.length];

        max[0] = min[0] = nums[0];
        int result = nums[0];

        for(int i=1; i<nums.length; i++){
            if(nums[i]>0){
                max[i]=Math.max(nums[i], max[i-1]*nums[i]);
                min[i]=Math.min(nums[i], min[i-1]*nums[i]);
            }else{
                max[i]=Math.max(nums[i], min[i-1]*nums[i]);
                min[i]=Math.min(nums[i], max[i-1]*nums[i]);
            }

            result = Math.max(result, max[i]);
        }

        return result;
    }

    public static void main(String[] args) {
        Scanner scan=new Scanner(System.in);
        System.out.println("Enter Array length: ");
        int l=scan.nextInt();
        int[] arr=new int[l];
        System.out.println("Enter Array:");
        for(int i=0; i<l; i++)
            arr[i]= scan.nextInt();

        System.out.println(maxProduct(arr));
    }
}

File: 02. Algorithms/01. Arrays/22. Max Product Subarray/Solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {

let maxCurrent = 1;
let minCurrent = 1;
let Max = nums[0];

for(let i = 0; i < nums.length; i++) {
    const temp = nums[i] * maxCurrent;

    maxCurrent = Math.max(temp, nums[i] * minCurrent, nums[i]);
    minCurrent = Math.min(temp, minCurrent * nums[i], nums[i]);

    Max = Math.max(Max, maxCurrent);
}

return Max;
};

